File: Engine/Source/Managed/ManagedCore/Modules/ModuleInfo.cs
using System;
using System.Collections.Generic;
using System.Reflection;

namespace Vanguard.ManagedCore.Modules
{
    class ModuleInfo
    {
        private String moduleName;
        public String ModuleName { get { return moduleName; } }

        private String moduleType;
        public String MduleType { get { return moduleType; } }

        private Type moduleInstanceType;
        public Type ModuleInstanceType { get { return moduleInstanceType; } }

        private IModule moduleInstance = null;
        public IModule ModuleInstance { get { return moduleInstance; } }

        public bool IsLoaded { get { return moduleInstance != null; } }

        public ModuleInfo(Type aInstanceType)
        {
            IModule tempInstance = (IModule)Activator.CreateInstance(aInstanceType);
            moduleName = tempInstance.GetModuleName();
            moduleType = tempInstance.GetModuleType();
            moduleInstanceType = aInstanceType;
            Console.WriteLine("Found managed module " + moduleName + " of type " + moduleType);
        }

        public void Load()
        {
            moduleInstance = (IModule)Activator.CreateInstance(moduleInstanceType);
            moduleInstance.LoadModule();
        }

        public void Unload()
        {
            if (moduleInstance != null)
                moduleInstance.UnloadModule();
            moduleInstance = null;
        }
    }
}

File: Engine/Source/Native/Core/Modules/ModuleInfo.cpp
#include "ModuleInfo.h"

namespace Vanguard
{
	typedef IModule * (*MODULE_INST_FUNCTION)();
	static const String InstantiateFunctionName = "InstantiateVanguardModule";

	ModuleInfo::ModuleInfo(const FilePath& aLibPath, const String& aName, const String& aType)
	{
		moduleName = aName;
		moduleType = aType;

		filePath = aLibPath;

		// Not loaded yet.
		dynamicLibReference = nullptr;
		moduleInstance = nullptr;
	}

	ModuleInfo::~ModuleInfo()
	{
		delete moduleInstance;
		delete dynamicLibReference;
	}

	ModuleInfo* ModuleInfo::LoadModuleAtPath(FilePath aModulePath)
	{
		DynamicLibrary* tempLoadedLib = new DynamicLibrary();

		if (!tempLoadedLib->Open(aModulePath))
		{
			// Could not load dynamic library
            DEBUG_ERROR("Could not load library " + aModulePath.GetFilename());
			delete tempLoadedLib;
			return nullptr;
		}

		MODULE_INST_FUNCTION InstantiationFunction = (MODULE_INST_FUNCTION)tempLoadedLib->GetFunction(InstantiateFunctionName.GetCharPointer());

		if (InstantiationFunction == nullptr)
		{
			// Could not find instantiation function (not a module library)
			delete tempLoadedLib;
			return nullptr;
		}

		IModule* tempModuleInstance = InstantiationFunction();

		String moduleName = tempModuleInstance->GetModuleName();
		String moduleType = tempModuleInstance->GetModuleType();

		// Got name and type, done with the temp stuff, so it can be unloaded.
		delete tempModuleInstance;
		delete tempLoadedLib;

		return new ModuleInfo(aModulePath, moduleName, moduleType);
	}

	void ModuleInfo::LoadModule()
	{
		if (!GetIsLoaded())
		{
			dynamicLibReference = new DynamicLibrary();

			if (!dynamicLibReference->Open(filePath))
			{
				// Could not load dynamic library
				UnloadModule();
				throw Exception(String("Could not load module " + filePath.GetFullPathName()).GetCharPointer());
			}

			MODULE_INST_FUNCTION InstantiationFunction = (MODULE_INST_FUNCTION)dynamicLibReference->GetFunction(InstantiateFunctionName.GetCharPointer());

			if (InstantiationFunction == nullptr)
			{
				throw Exception(String("Could not load module " + filePath.GetFullPathName()).GetCharPointer());
			}

			Log::Message("Loading native module " + moduleName, "Modules");

			moduleInstance = InstantiationFunction();
			moduleInstance->LoadModule();
		}
	}

	void ModuleInfo::UnloadModule()
	{
		if (moduleInstance)
		{
			moduleInstance->UnloadModule();
			delete moduleInstance;
			moduleInstance = nullptr;
		}

		if (dynamicLibReference)
		{
			dynamicLibReference->Close();
			delete dynamicLibReference;
			dynamicLibReference = nullptr;
		}
	}
}
